### DISCUSSION TITLE
Working with auth0 in Vercel preview environments

### DISCUSSION THREAD

**Discussion by jacksonblankenship** [vercel/vercel#4811] [Help] [answered]
I'm having some trouble configuring auth0 to play nice with Vercel previews. I need to set two environment variables that are URL dependent. For example, when running on localhost the following variables evaluate to the localhost address. When running in production, the variables evaluate to my purchased domain name. Both of these configurations are working as intended.

```js
REDIRECT_URI: process.env.REDIRECT_URI || 'http://localhost:3000/api/callback',
POST_LOGOUT_REDIRECT_URI: process.env.POST_LOGOUT_REDIRECT_URI || 'http://localhost:3000/',
```

The issue I'm facing is the middle ground: *vercel previews*. They have very unpredictable URLs which means I can't whitelist them in my auth0 settings, specifically Allowed Callback URLs and Allowed Logout URLs. This ultimately causes the following error in my auth0 logs...

```Error
Callback URL mismatch. https://my-fake-app-5dsd7jxul.vercel.app/api/callback is not in the list of allowed callback URLs
```

I was curious to see if any of you have found a clever way to get around this issue when working with auth0 in preview environments. I can't seem to find anyone who's documented a safe solution.

**Comment by jonahallibone** [comment]
@jacksonblankenship Did you ever get a response on this? I am in the same boat. I have large preview branches but they are only available locally for this reason

**Comment by mcsdevv** [comment, marked as answer]
Hello @jacksonblankenship @jonahallibone, since this was posted, Auth0 has released wildcard URL support for subdomains making this much easier to work with: https://auth0.com/docs/applications/wildcards-for-subdomains

**Reply by catedm** [reply]
@jacksonblankenship - i am still having difficulty getting my auth0 login flow to work with vercel preview urls. after authentication - auth0 always redirects me to my production domain instead of a preview url domain. i have added ` https://*.vercel.app` to the list of Allowed Callback URLs in my auth0 configuration.

Do you have any extra tips you can give me to help get this working? Thank you in advance.

**Reply by jacksonblankenship** [reply]
Hey @catedm, I've since moved away from Auth0 and into NextAuth so I don't have any additional advice. Sorry, good luck!

**Reply by andrevenancio** [reply]
@catedm try this: https://github.com/auth0/nextjs-auth0/blob/main/examples/README.md#wildcards

**Reply by theborowski** [reply]
Updated URL for wildcards: https://github.com/auth0/nextjs-auth0/blob/main/example-app/README.md#wildcards
